### PR TITLE
Feature/bakta flag

### PR DESCRIPTION
**Pull request by tgurbich**
Added Bakta to the pipeline - if the user runs the pipeline with a --bakta flag, bacterial genomes will be annotated with Bakta instead of Prokka

### CODE REVIEW

File: workflows/mettannotator.nf [hunk 1]
```diff
@@ -164,12 +175,25 @@ workflow METTANNOTATOR {
 
     LOOKUP_KINGDOM( assemblies )
 
-    PROKKA( assemblies.join( LOOKUP_KINGDOM.out.detected_kingdom ))
+    gene_caller = channel.empty()
+
+    if ( params.bakta && val(cat LOOKUP_KINGDOM.out.detected_kingdom) == "Bacteria" ) {
```
Comment (mberacochea) [inline workflows/mettannotator.nf:180]: This is new to me `val(cat LOOKUP_KINGDOM.out.detected_kingdom)`. Does it work?

The lookup is per assembly, so this won't work when mixed (arche and bacteria) exist in the sample sheet.

I think this is what you need:
```nextflow
if (params.bakta ) {
   // I'm not sure if 3 is the right index for the branch - it[] thingy
   assemblies.join( LOOKUP_KINGDOM.out.detected_kingdom ).branch {
       bacteria: it[3] == "Bacteria"
       archea: it[3] == "Archea"
   }.set { assemblies_to_annotate }

   BAKTA_BATKA( assemblies_to_annotate.bacteria, bakta_db )

   PROKKA( assemblies_to_annotate.archea )

   // I'm not sure if this is going to work downstream
   gene_caller = BAKTA_BATKA.out.mix( PROKKA.out )

   ch_versions = ch_versions.mix(BAKTA_BATKA.out.versions.first())
   ch_versions = ch_versions.mix(PROKKA.out.versions.first())

} else {

   PROKKA( assemblies.join( LOOKUP_KINGDOM.out.detected_kingdom ) )

   ch_versions = ch_versions.mix(PROKKA.out.versions.first())

   gene_caller = PROKKA
}
```

**Review by mberacochea — COMMENTED** [review]
I know this is not ready.. but I saw the email and gave it a read.

File: workflows/mettannotator.nf (hunk 1 above)
Comment (mberacochea) [reply workflows/mettannotator.nf]: second iteration:
```nextflow
annotations_fna = channel.empty()
annotations_gbk = channel.empty()
annotations_faa = channel.empty()
annotations_gff = channel.empty()

if (params.bakta ) {
   // I'm not sure if 3 is the right index for the branch - it[] thingy
   assemblies.join( LOOKUP_KINGDOM.out.detected_kingdom ).branch {
       bacteria: it[3] == "Bacteria"
       archea: it[3] == "Archea"
   }.set { assemblies_to_annotate }

   BAKTA_BATKA( assemblies_to_annotate.bacteria, bakta_db )

   PROKKA( assemblies_to_annotate.archea )

   ch_versions = ch_versions.mix(BAKTA_BATKA.out.versions.first())
   ch_versions = ch_versions.mix(PROKKA.out.versions.first())

   annotations_fna = annotations_fna.mix( BAKTA_BATKA.out.fna ).mix( PROKKA.out.fna )
   annotations_gbk = annotations_gbk.mix( BAKTA_BATKA.out.gbk ).mix( PROKKA.out.gbk )
   annotations_faa = annotations_faa.mix( BAKTA_BATKA.out.faa ).mix( PROKKA.out.faa )
   annotations_gff = annotations_gff.mix( BAKTA_BATKA.out.gff ).mix( PROKKA.out.gff )

} else {

   PROKKA( assemblies.join( LOOKUP_KINGDOM.out.detected_kingdom ) )

   ch_versions = ch_versions.mix(PROKKA.out.versions.first())

   annotations_fna = PROKKA.out.fna
   annotations_gbk = PROKKA.out.gbk
   annotations_faa = PROKKA.out.faa
   annotations_gff = PROKKA.out.gff
}
```

Comment (mberacochea) [reply workflows/mettannotator.nf]: This requires replacing the gene_caller.out downstream

File: modules/local/bakta_getdb.nf [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    wget -r -nH --cut-dirs=5 \\
+    ftp://ftp.ncbi.nlm.nih.gov/pathogen/Antimicrobial_resistance/AMRFinderPlus/database/3.12/2024-01-31.1/
+    rm -r bakta/amrfinderplus-db/*
+    mv 2024-01-31.1 bakta/amrfinderplus-db/
+
+    amrfinder_index bakta/amrfinderplus-db/2024-01-31.1
+
+    mv bakta/amrfinderplus-db/2024-01-31.1/ bakta/amrfinderplus-db/latest/
+
+    echo "2024-01-19" > bakta/Version.txt
```
Comment (mberacochea) [inline modules/local/bakta_getdb.nf:28]:
```suggestion
    echo "2024-01-19" > bakta/VERSION.txt
```

File: modules/local/sanntis.nf [hunk 3]
```diff
@@ -15,14 +15,16 @@ process SANNTIS {
     path "versions.yml"                   , emit: versions
 
     script:
+    def genbank_extension = prokka_gbk.extension
     if (interproscan_tsv.extension == "gz") {
         """
         gunzip -c ${interproscan_tsv} > interproscan.tsv
+        grep -v "/protein_id=" ${prokka_gbk} > ${meta.prefix}_prepped.${genbank_extension}
```
Comment (mberacochea) [inline modules/local/sanntis.nf:22]: This is because `SanntiS` doesn't like that bit on the gbk file?

File: workflows/mettannotator.nf [hunk 4]
```diff
@@ -156,20 +161,66 @@ workflow METTANNOTATOR {
...
     LOOKUP_KINGDOM( assemblies )
 
-    PROKKA( assemblies.join( LOOKUP_KINGDOM.out.detected_kingdom ))
+    assembliesWithKingdom = assemblies.join( LOOKUP_KINGDOM.out.detected_kingdom ).map{ meta, file1, file2 -> {
+        def parts = file2.toString().split('/')
+        def fileName = parts[-1]
+        def nameParts = fileName.split('_')
+        def kingdomName = nameParts[0]
+        return [meta, file1, kingdomName]
+        }
```
Comment (mberacochea) [inline workflows/mettannotator.nf:188]: I'm not a fan of this solution, let me try a different approach.

**Review by mberacochea — COMMENTED** [review]
Good stuff.. I left a couple of minor comments. And I'm trying to tidy up the kingdom stuff, using the file name is a brittle and likely to cause problems in the future.

File: modules/local/sanntis.nf (hunk 3 above)
Comment (tgurbich) [reply modules/local/sanntis.nf:22]: Basically yes - it preferentially takes the protein name from the protein_id field but in Bakta the format of it is different and causes SanntiS to output an empty file. With that line removed, it uses the locus tag to get the name and that works well. I reported it but using this as a temporary solution.

File: modules/local/sanntis.nf [hunk 5]
```diff
@@ -15,14 +15,16 @@ process SANNTIS {
     path "versions.yml"                   , emit: versions
 
     script:
+    def genbank_extension = prokka_gbk.extension
     if (interproscan_tsv.extension == "gz") {
         """
         gunzip -c ${interproscan_tsv} > interproscan.tsv
```
Comment (mberacochea) [inline modules/local/sanntis.nf:21]:
```suggestion
        gunzip -c ${interproscan_tsv} > interproscan.tsv
        // Workaround implemented for SanntiS due to discrepancies in Bakta's format, resulting in empty output files.
```

**Review by mberacochea — APPROVED** [review]

File: modules/local/bakta_getdb.nf (hunk 2 above)
Comment (mberacochea) [reply modules/local/bakta_getdb.nf:28]: I'll adjust this one with the reworked `assembliesWithKingdom` thingy

**Pull request merged**